thewalrus/internal_modes/pnr_statistics.py

```diff
@@ -120,3 +126,121 @@ def pnr_prob(covs, i, hbar=2):
+
+    Args:
+        A (array): input matrix
+        blocks (list): how to block (coarse graining) different outcomes
+        repeats (list): pattern for repetition but with one added in each repetition
+
+    Returns:
+        value of the hafnian
+    """
+    # Note that the two lines below cannot be done inside numba hence the need for this function
+    repeats = tuple(val + 1 for val in repeats)
+    blocks = [np.array(val, dtype=np.int32) for val in blocks]
+    return _haf_blocked_numba(A, blocks, repeats)
+
+
+@numba.jit(nopython=True)
+def _haf_blocked_numba(A, blocks, repeats_p):
+    """Calculates the hafnian of the matrix with a given block and repetition pattern.
+
+    Args:
+        A (array): input matrix
+        blocks (list): how to block (coarse graining) different outcomes
+        repeats_p (list): pattern for repetition but with one added in each repetition
+
+    Returns:
+        value of the hafnian
+    """
+    n = sum(repeats_p) - len(repeats_p)
+    num_indices = 0
+    for block in blocks:
+        num_indices += len(block)
+    netsum = 0.0 + 0j
+    coeff_vect = np.zeros(num_indices, dtype=np.int32)
+    for index in np.ndindex(*repeats_p):
+        coeff_pref = 1.0 + 0j
+        for i, block in enumerate(blocks):
+            (coeff, val) = finite_difference_operator_coeffs(repeats_p[i] - 1, index[i])
+            coeff_pref *= coeff
+            for mode in block:
+                coeff_vect[mode] = val
+        AX_S = get_AX_S(coeff_vect, A)
+
+        netsum += coeff_pref * f_from_matrix(AX_S, 2 * n)[n]
+    return netsum
+
+
+def probabilities_single_mode(cov, pattern, normalize=False, LO_overlap=None, cutoff=13, hbar=2):
+    """
+    Calculates the diagonal of the density matrix, hence the name probabilities, of first mode when heralded by pattern on a zero-displaced, M-mode Gaussian state
+    where each mode contains K internal modes.
+
+    Args:
+        cov (array): 2MK x 2MK covariance matrix
+        pattern (dict): heralding pattern total photon number in the spatial modes (int), indexed by spatial mode
+        normalize (bool): whether to normalise the output density matrix
+        LO_overlap (array): overlap between internal modes and local oscillator
+        cutoff (int): photon number cutoff. Should be odd. Even numbers will be rounded up to an odd number
+        hbar (float): the value of hbar (default 2)
+    Returns:
+        array[complex]: (cutoff+1, cutoff+1) dimension density matrix
+    """
+
+    # The lines until A =  Amat(...) are copies from density_single_mode
+    cov = np.array(cov).astype(np.float64)
+    M = len(pattern) + 1
+    K = cov.shape[0] // (2 * M)
+    if not set(list(pattern.keys())).issubset(set(list(np.arange(M)))):
+        raise ValueError("Keys of pattern must correspond to all but one spatial mode")
+    N_nums = np.array(list(pattern.values()))
```

Comment:
```suggestion
    N_nums = list(pattern.values())
```
For use in `haf_blocked` (see comment on line 237)